rts/motoko-rts/src/gc/mark_compact/bitmap.rs

```diff
@@ -85,41 +159,42 @@ pub unsafe fn iter_bits() -> BitmapIter {
 //
 // (We actually need less bits than that as when the heap is full we can't allocate bitmap and mark
 // stack and can't do GC)
-pub const BITMAP_ITER_END: u32 = 1024 * 1024 * 1024;
+pub const BITMAP_ITER_END: u32 = u32::MAX;
 
 impl BitmapIter {
     /// Returns the next bit, or `BITMAP_ITER_END` if there are no more bits set.
     pub fn next(&mut self) -> u32 {
-        debug_assert!(self.current_word_idx <= self.size);
+        debug_assert!(self.current_bit_idx <= self.size);
+
+        if self.current_bit_idx == self.size {
+            return BITMAP_ITER_END;
+        }
 
         // Outer loop iterates 64-bit words
         loop {
-            if self.current_word == 0 && self.current_word_idx == self.size {
-                return BITMAP_ITER_END;
-            }
-
             // Inner loop iterates bits in the current word
             while self.current_word != 0 {
-                if self.current_word & 0b1 == 0b1 {
-                    let bit_idx = (self.current_word_idx * 64) + (64 - self.bits_left);
+                if self.current_word & 0b1 != 0 {
+                    let bit_idx = self.current_bit_idx;
                     self.current_word >>= 1;
-                    self.bits_left -= 1;
+                    self.current_bit_idx += 1;
                     return bit_idx;
                 } else {
                     let shift_amt = self.current_word.trailing_zeros();
                     self.current_word >>= shift_amt;
-                    self.bits_left -= shift_amt;
+                    self.current_bit_idx += shift_amt;
                 }
             }
 
-            // Move on to next word
-            self.current_word_idx += 1;
-            if self.current_word_idx == self.size {
+            // Move on to next word (always 64-bit boundary)
+            self.current_bit_idx += self.leading_zeros;
+            if self.current_bit_idx == self.size {
                 return BITMAP_ITER_END;
             }
-            self.current_word =
-                unsafe { *(BITMAP_PTR as *const u64).add(self.current_word_idx as usize) };
-            self.bits_left = 64;
+            self.current_word = unsafe {
+                *(BITMAP_FORBIDDEN_PTR.add(self.current_bit_idx as usize / 8) as *const u64)
+            };
+            self.leading_zeros = self.current_word.leading_zeros();
```

Comment:
I'm a bit confused about why we need a state (`self.leading_zeros`) here. When we move on to the next word, we need to update the bit index with `self.current_bit_idx += self.current_word.leading_zeros()`. After this the leading zeros won't be needed anymore so the state can be removed. Am I missing anything?

It also seems to me like the code above is buggy, because it updates `current_bit_idx` *before* moving on to the next word.

Reply:
`self.leading_zeros` is needed because inner loop doesn't handle the leading zeros of the `self.current_word`. So when the topmost bit is _not_ set, then `self.current_bit_idx` won't be incremented to the next 64-bit boundary _by the loop_ and we have to do it separately. Also if the inner loop never gets entered (`self.current_word == 0`) then we have to jump to the next word boundary by adding 64 bits. In this case `self.current_bit_idx += self.leading_zeros` will do the right thing, because `self.leading_zeros == 64`, and we can make progress.

`self.leading_zeros` needs to be in the `self` state (and recorded before the inner loop), because the shifting of `self.current_word` would increase the leading zeros in it.

I don't think we have a bug here, hopefully the above makes it clear why.

Reply:
I found a bug while working on simplifying this method. When `self.leading_zeros` is 0, this line

```rust
            self.current_bit_idx += self.leading_zeros;
```

does not increment current bit idx, and cause the iterator to yield the same bit index twice.

I'm surprised that this isn't causing heap corruption. I realized this case in the RTS tests, but the tests somehow pass..

Reply:
We should be able to easily extend `test_bit_iter` test function to check for visited bits, and fail if the iterator yields the same bit multiple times. @ggreif do you want to do this? I can do this if you prefer that. We can then merge it, merge master to this branch, and make sure we won't yield same bit index multiple times in this branch.

Reply:
I wonder how that could happen, let's analyse! When entering `BitmapIter::next` we have 2 cases:
- `self.leading_zeros` is 0 (implying `current_word` is `0x[8-F].......`)
- `self.leading_zeros` is `> 0` (implying `current_word` is `0x[0-7].......`)
In the first case we enter the inner loop and find a bit:
- either the LSBit -> increments the `self.current_bit_idx` by one
- a string of trailing zeros -> increments the `self.current_bit_idx` by at least one
In the other case we don't enter the loop and `self.current_bit_idx` gets incremented by nonzero `self.leading_zeros`.

I still can't see a bug.

Reply:
OK, that makes sense. Thanks for the explanation.